Commit №13.1:

sashajaaa opened the pull request:
feat:
-Создан пакет booking, содержащий файлы классов BookingDto, BookingMapper, Booking и BookingController; -Создан пакет item, содержащий файлы классов ItemDto, ItemMapper, Item, ItemRepositoryImpl, ItemService, ItemController и файл интерфейса ItemRepository; -Создан пакет request, содержащий файлы классов ItemRequestDto, ItemRequestMapper, ItemRequest и ItemRequestController; -Создан пакет user, содержащий файлы классов UserDto, UserMapper, User, UserRepositoryImpl и файл интерфейса UserRepository; -Создан пакет exception, содержащий файлы классов ItemNotFoundException, UserAlreadyExistsException, UserNotFoundException и ValidationException; -Создан пакет handler, содержащий файл класса ErrorHandler.

sashajaaa committed:
Commit №13.1:
feat:
-Создан пакет booking, содержащий файлы классов BookingDto, BookingMapper, Booking и BookingController;
-Создан пакет item, содержащий файлы классов ItemDto, ItemMapper, Item, ItemRepositoryImpl, ItemService, ItemController и файл интерфейса ItemRepository;
-Создан пакет request, содержащий файлы классов ItemRequestDto, ItemRequestMapper, ItemRequest и ItemRequestController;
-Создан пакет user, содержащий файлы классов UserDto, UserMapper, User, UserRepositoryImpl и файл интерфейса UserRepository;
-Создан пакет exception, содержащий файлы классов ItemNotFoundException, UserAlreadyExistsException, UserNotFoundException и ValidationException;
-Создан пакет handler, содержащий файл класса ErrorHandler.

kalitva commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,16 @@
+import ru.practicum.shareit.booking.model.Booking;
+
+public class BookingMapper {
+    public static BookingDto toBookingDto(Booking booking) {
```
Вместо конструкторов для создания дтошек лучше используй `@Builder` ломбока

https://projectlombok.org/features/Builder

kalitva replied:
Помечай комменты как resolved, где-то под комментом у тебя должна быть такая галочка

kalitva commented on src/main/java/ru/practicum/shareit/exception/ItemNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+public class ItemNotFoundException extends IllegalArgumentException {
+    public ItemNotFoundException(String message) {
+        super(message);
+        log.error(message);
```
Конструктор исключения не лучшее место для какого-либо логирования, логируй исключение там где его перехватываешь

kalitva commented on src/main/java/ru/practicum/shareit/handler/ErrorHandler.java:
```diff
@@ -0,0 +1,50 @@
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ErrorResponse handleItemNotFoundException(final ItemNotFoundException e) {
+        return new ErrorResponse(e.getMessage());
+    }
```
Предлагаю эти два обработчика, возвращающих 404, объединить в один, смотри пример в доках:

```
@ExceptionHandler({FileSystemException.class, RemoteException.class})
public ResponseEntity<String> handle(IOException ex) {
    // ...
}
```

https://docs.spring.io/spring-framework/docs/current/reference/html/web.html#mvc-ann-exceptionhandler

kalitva commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,73 @@
 @RequestMapping("/items")
 public class ItemController {
-}
+    private static final String OWNER = "X-Sharer-User-Id";
```
Лучше `OWNER --> OWNER_ID_HEADER`

kalitva commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+import static java.util.stream.Collectors.toList;
+
+@Component("ItemRepositoryImpl")
```
`@Component --> @Repository`

https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-stereotype-annotations

kalitva commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -0,0 +1,88 @@
+        }
+            newItemDto = mapper.toItemDto(itemRepository.update(mapper.toItem(itemDto, ownerId)));
+        return newItemDto;
+        }
```
Поправь здесь форматирование, съехали отступы

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,102 @@
+    public User create(User user) {
+        if (users.values()
+                .stream()
+                .noneMatch(u -> u.getEmail().equals(user.getEmail()))) {
```
1. Вынеси результат стрима в переменную, и используй ее в if, так не очень удобно читается
2. Все такого рода проверки должны быть в сервисах, репозиторий для того чтобы достать, сохранить и удалить, не более

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public User update(User user) {
+        if (user.getId() == null) {
```
В сервисе

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,102 @@
+        if (!users.containsKey(user.getId())) {
+            throw new UserNotFoundException("User with ID = " + user.getId() + " not found.");
+        }
+        if (user.getName() == null) {
```
Это тоже

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public User delete(Long userId) {
+        if (userId == null) {
```
Проверяй наличие объекта в сервисе, вызывая метод реаозитория

kalitva commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,119 @@
+        if (!items.containsKey(item.getId())) {
+            throw new ItemNotFoundException("Item with ID = " + item.getId() + " not found.");
+        }
+        if (item.getName() == null) {
```
Это все делается в сервисе, не в репозитории

sashajaaa committed:
Commit №13.1:
refactor:
-исключения UserNotFoundException и ItemNotFoundException объеденины в одно;
-переработан класс BookingDto;
-переработан класс BookingMapper;
-переработан класс ErrorHandler;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан интерфейс ItemRepository;
-переработан класс ItemRepositoryImpl;
-переработан класс ItemService;
-переработан класс ItemController;
-переработан класс ItemRequestDto;
-переработан класс ItemRequestMapper;
-переработан класс ItemRequest;
-переработан класс UserDto;
-переработан класс UserMapper;
-переработан класс User;
-переработан интерфейс UserRepository;
-переработан класс UserRepositoryImpl;
-переработан класс UserService.

kalitva commented on src/main/java/ru/practicum/shareit/exception/AlreadyExistsException.java:
```diff
@@ -0,0 +1,11 @@
+public class AlreadyExistsException extends IllegalArgumentException {
+    public AlreadyExistsException(String message) {
+        super(message);
+        log.error(message);
```
Не самая лучшая идея логировать в конструкторе исключения

kalitva commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,89 @@
+        if (item.getName() == null) {
+            item.setName(items.get(item.getId()).getName());
+        }
+        if (item.getDescription() == null) {
```
Пожалуйста, делай эти проверки в сервисе, а не в репозитории

kalitva commented on src/main/java/ru/practicum/shareit/item/repository/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,89 @@
+                .filter(item -> item.getOwnerId().equals(ownerId))
+                .map(Item::getId)
+                .collect(toList()));
+        for (Long deleteId : deleteIds) {
```
Вместо `collect` поставь `forEach` и проитерийруйся по элемента, здесь нет необходимости собирать в список

kalitva commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public ItemDto update(ItemDto itemDto, Long ownerId, Long itemId) {
+        ItemDto newItemDto;
```
Не вижу причин зачем объявлять отдельно, объяви там же, где и инициализируешь (на 52 линии)

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+        if (user.getName() == null) {
+            user.setName(users.get(user.getId()).getName());
+        }
+        if (user.getEmail() == null) {
```
Эти проверки в сервис

kalitva commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,76 @@
+@Qualifier("UserRepositoryImpl")
+public class UserRepositoryImpl implements UserRepository {
+
+    public Map<Long, User> users;
```
private

sashajaaa committed:
Commit №13.3:
refactor:
-переработан класс AlreadyExistsException;
-переработан класс ItemRepositoryImpl;
-переработан класс ItemService;
-переработан класс UserRepositoryImpl;
-переработан класс UserService.

sashajaaa committed:
Commit №13.3:
refactor:
-переработан класс AlreadyExistsException;
-переработан класс ItemRepositoryImpl;
-переработан класс ItemService;
-переработан класс UserRepositoryImpl;
-переработан класс UserService.
